src/material-experimental/mdc-chips/grid-key-manager.ts

```diff
@@ -0,0 +1,272 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {QueryList} from '@angular/core';
+import {Subject} from 'rxjs';
+import {
+  UP_ARROW,
+  DOWN_ARROW,
+  LEFT_ARROW,
+  RIGHT_ARROW,
+  TAB,
+} from '@angular/cdk/keycodes';
+
+
+/** The keys handled by the GridKeyManager keydown method. */
+export const NAVIGATION_KEYS = [TAB, DOWN_ARROW, UP_ARROW, RIGHT_ARROW, LEFT_ARROW];
+
+/** This interface is for cells that make up a GridKeyManagerRow. */
+export interface GridKeyManagerCell {}
+
+/** This interface is for rows that can be passed to a GridKeyManager. */
+export interface GridKeyManagerRow<T extends GridKeyManagerCell> {
+  cells: T[];
+}
+
+/**
+ * This class manages keyboard events for grids. If you pass it a query list
+ * of GridKeyManagerRow, it will set the active cell correctly when arrow events occur.
+ *
+ * GridKeyManager expects that rows may change dynamically, but the cells for a given row are
+ * static. It also expects that all rows have the same number of cells.
+ */
+export class GridKeyManager<T extends GridKeyManagerCell> {
+  private _activeRowIndex = -1;
+  private _activeColumnIndex = -1;
+  private _activeRow: GridKeyManagerRow<T> | null = null;
+  private _activeCell: T | null = null;
+  private _horizontal: 'ltr' | 'rtl' = 'ltr';
+
+  constructor(private _rows: QueryList<GridKeyManagerRow<T>> | GridKeyManagerRow<T>[]) {
+    // We allow for the rows to be an array because, in some cases, the consumer may
+    // not have access to a QueryList of the rows they want to manage (e.g. when the
+    // rows aren't being collected via `ViewChildren` or `ContentChildren`).
+    if (_rows instanceof QueryList) {
+      _rows.changes.subscribe((newRows: QueryList<GridKeyManagerRow<T>>) => {
+        if (this._activeRow) {
+          const newIndex = newRows.toArray().indexOf(this._activeRow);
+
+          if (newIndex > -1 && newIndex !== this._activeRowIndex) {
+            this._activeRowIndex = newIndex;
+          }
+        }
+      });
+    }
+  }
+
+  /**
+   * Stream that emits any time the TAB key is pressed, so components can react
+   * when focus is shifted off of the grid.
+   */
+  tabOut: Subject<void> = new Subject<void>();
```

Comment:
I'd prefer to omit the `tabOut` API from this class. I consider it a mistake that we added it to `ListKeyManager`, since it's not related to any particular interaction pattern, it's universal for all controls.